### PR TITLE
fix(gnovm): Support mixed key and unkeyed elements in a slice

### LINKED ISSUE #3464: mixed keyed and un-keyed elements elements in slice are not supported
### Description

Mixed keyed and unkeyed elements in a slice should be allowed.

According to the go specs https://go.dev/ref/spec

```
For array and slice literals the following rules apply:
- Each element has an associated integer index marking its position in the array.
- An element without a key uses the previous element's index plus one. If the first element has no key, its index is zero.
```

### Your environment

* Go version (example: go1.22)
* OS and CPU architecture (MacOS/ARM64)
* Gno commit hash causing the issue (6472eea9bb1b460471941d6f8e028ce309a5c2b1)

### Steps to reproduce

```
// slice4.gno
package main

func main() {
	var x = []int{4: 14, 25, 1: 33, 34}
	println(len(x), x[2])
}

// Output:
//  6 34
```

 - add above file to gnovm/tests/files/slice/slice4.gno
 - go test gnovm/pkg/gnolang/files_test.go -run 'TestFiles/slice/'

### Expected behaviour

It should pass, and the output should be `6 34`

https://go.dev/play/p/QWy6BXtDJZV?v=goprev

### Actual behaviour

--- FAIL: TestFiles (0.04s)
    --- FAIL: TestFiles/slice/slice4.gno (0.00s)
panic: slice composite literal cannot mix keyed and unkeyed elements [recovered]
	panic: slice composite literal cannot mix keyed and unkeyed elements [recovered]
	panic: interface conversion: gnolang.Stmt is *gnolang.ValueDecl, not *gnolang.bodyStmt [recovered]
	panic: interface conversion: gnolang.Stmt is *gnolang.ValueDecl, not *gnolang.bodyStmt

### PR DESCRIPTION
**Pull request by aronpark1007**
## Description

closes #3464

GnoVM was panicking on mixed keyed/unkeyed slice composite literals (e.g. `[]int{4: 14, 25, 1: 33, 34}`), which is valid Go syntax. All slice literals are now routed through `OpSliceLit2`, which distinguishes keyed and unkeyed elements. Unkeyed elements automatically get the previous `index +1`. Also, `maxVal` and `idx` are initialized to -1 so that empty slice literals (e.g. `[]rune{}`) produce a zero-length slice correctly.

### CODE REVIEW

**Comment by Gno2D2** [comment]
#### 🛠 PR Checks Summary
All **Automated Checks** passed. ✅

##### Manual Checks (for Reviewers):
- [ ] **IGNORE** the bot requirements for this PR (force green CI check)
- [ ] The pull request description provides enough details


<details><summary>Read More</summary>

🤖 This bot helps streamline PR reviews by verifying automated checks and providing guidance for contributors and reviewers.

##### ✅ Automated Checks (for Contributors):
 🟢 Maintainers must be able to edit this pull request ([more info](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
 🟢 Pending initial approval by a review team member, or review from tech-staff


##### ☑️ Contributor Actions:
1. Fix any issues flagged by automated checks.
2. Follow the Contributor Checklist to ensure your PR is ready for review.
	- Add new tests, or document why they are unnecessary.
	- Provide clear examples/screenshots, if necessary.
	- Update documentation, if required.
	- Ensure no breaking changes, or include `BREAKING CHANGE` notes.
	- Link related issues/PRs, where applicable.

##### ☑️ Reviewer Actions:
1. Complete manual checks for the PR, including the guidelines and additional checks if applicable.

##### 📚 Resources:
- [Report a bug with the bot](https://www.github.com/gnolang/gno/issues/3238).
- [View the bot’s configuration file](https://github.com/gnolang/gno/tree/master/contribs/github-bot/internal/config/config.go).

<details><summary><b>Debug</b></summary><blockquote>
<details><summary><b>Automated Checks</b></summary><blockquote>

<details><summary>Maintainers must be able to edit this pull request (more info)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 The pull request was created from a fork (head branch repo: aronpark1007/gno)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 Maintainer can modify this pull request

```
</blockquote></details>

<details><summary>Pending initial approval by a review team member, or review from tech-staff</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 The base branch matches this pattern: ^master$
    └── 🟢 Not (🔴 Pull request author is a member of the team: tech-staff)

```
### Then
```
🟢 Requirement satisfied
└── 🟢 If
    ├── 🟢 Condition
    │   └── 🟢 Or
    │       ├── 🔴 At least one of these user(s) reviewed the pull request: [aronpark1007 davd-gzl jefft0 notJoon omarsy MikaelVallenet] (with state "APPROVED")
    │       ├── 🟢 At least 1 user(s) of the team tech-staff reviewed pull request
    │       └── 🔴 This pull request is a draft
    └── 🟢 Then
        └── 🟢 Not (🔴 This label is applied to pull request: review/triage-pending)

```
</blockquote></details>

</blockquote></details>


<details><summary><b>Manual Checks</b></summary><blockquote>

<details><summary>**IGNORE** the bot requirements for this PR (force green CI check)</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 On every pull request

```
### Can be checked by

- Any user with comment edit permission

</blockquote></details>

<details><summary>The pull request description provides enough details</summary><blockquote>

### If
```
🟢 Condition met
└── 🟢 And
    ├── 🟢 Not (🔴 Pull request author is a member of the team: core-contributors)
    └── 🟢 Not (🔴 Pull request author is user: dependabot[bot])

```
### Can be checked by
 - team core-contributors

</blockquote></details>

</blockquote></details>

</blockquote></details>

</details>

File: gnovm/pkg/gnolang/op_expressions.go [hunk 1]
```diff
@@ -356,28 +356,15 @@ func (m *Machine) doOpCompositeLit() {
...
-					panic("slice composite literal cannot mix keyed and unkeyed elements")
-				}
-				m.PushExpr(x.Elts[i].Value)
-				m.PushOp(OpEval)
+		m.PushOp(OpSliceLit2)
+		// evaluate item values; also evaluate key if present
+		for i := len(x.Elts) - 1; 0 <= i; i-- {
+			m.PushExpr(x.Elts[i].Value)
+			m.PushOp(OpEval)
+			if x.Elts[i].Key != nil {
```
Comment (notJoon) [inline gnovm/pkg/gnolang/op_expressions.go]: the current implementation consolidates all slice literals into `OpSliceLit2`, but looking at it now, it seems like unnecessary performance overhead could occur in the all-unkeyed case.

maybe we can do this something like this:

```go
hasKey := false
for _, elt := range x.Elts {
	if elt.Key != nil {
		hasKey = true
		break
	}
} 
if hasKey {
	m.PushOp(OpSliceLi2)
	// eval keys + values
} else {
	m.PushOp(OpSliceLit)
	// eval values only
}
```

This keeps the unkeyed fast path and utilizes `OpSliceLit` as-is.

File: gnovm/tests/files/slice6.gno [hunk 2]
```diff
@@ -0,0 +1,10 @@
+// slice4.gno
```
Comment (notJoon) [inline gnovm/tests/files/slice6.gno]:
```suggestion
```

File: gnovm/tests/files/slice6.gno
Comment (notJoon) [inline gnovm/tests/files/slice6.gno:1]: this file has not been formatted.

**Review by notJoon — CHANGES_REQUESTED** [review]

File: gnovm/pkg/gnolang/op_expressions.go (hunk 1 above)
Comment (ltzmaxwell) [inline gnovm/pkg/gnolang/op_expressions.go]: do we still need `OpSliceLit` after this?

Comment (aronpark1007) [reply gnovm/pkg/gnolang/op_expressions.go]: Sharing benchmark results comparing three approaches for slice composite literal handling, including the current main branch.

---

### Approaches

**1. `doOpCompositeLitCheckFirst` (main branch)**
Branches on keyed/unkeyed by inspecting only the first element; includes a panic guard inside the loop.
Note: the current version does not support mixed slices — included here purely as a performance baseline for keyed/unkeyed cases.

```go
case *SliceType:
    if len(x.Elts) > 0 && x.Elts[0].Key != nil {
        m.PushOp(OpSliceLit2)
        for i := len(x.Elts) - 1; 0 <= i; i-- {
            if x.Elts[i].Key == nil {
                panic("slice composite literal cannot mix keyed and unkeyed elements")
            }
            m.PushExpr(x.Elts[i].Value)
            m.PushOp(OpEval)
            m.PushExpr(x.Elts[i].Key)
            m.PushOp(OpEval)
        }
    } else {
        m.PushOp(OpSliceLit)
        for i := len(x.Elts) - 1; 0 <= i; i-- {
            if x.Elts[i].Key != nil {
                panic("slice composite literal cannot mix keyed and unkeyed elements")
            }
            m.PushExpr(x.Elts[i].Value)
            m.PushOp(OpEval)
        }
    }
```

**2. `doOpCompositeLitSinglePass`**
Always uses `OpSliceLit2`; conditionally pushes the key inside the loop only when present.

```go
case *SliceType:
    m.PushOp(OpSliceLit2)
    for i := len(x.Elts) - 1; 0 <= i; i-- {
        m.PushExpr(x.Elts[i].Value)
        m.PushOp(OpEval)
        if x.Elts[i].Key != nil {
            m.PushExpr(x.Elts[i].Key)
            m.PushOp(OpEval)
        }
    }
```

**3. `doOpCompositeLitPrescan`**
Pre-scans all elements (breaking on first key found) to decide the branch. Also handles mixed input without panicking.

```go
case *SliceType:
    hasKey := false
    for _, elt := range x.Elts {
        if elt.Key != nil {
            hasKey = true
            break
        }
    }
    if hasKey {
        m.PushOp(OpSliceLit2)
        for i := len(x.Elts) - 1; 0 <= i; i-- {
            m.PushExpr(x.Elts[i].Value)
            m.PushOp(OpEval)
            if x.Elts[i].Key != nil {
                m.PushExpr(x.Elts[i].Key)
                m.PushOp(OpEval)
            }
        }
    } else {
        m.PushOp(OpSliceLit)
        for i := len(x.Elts) - 1; 0 <= i; i-- {
            m.PushExpr(x.Elts[i].Value)
            m.PushOp(OpEval)
        }
    }
```

---

### Benchmark Conditions

- **unkeyed / keyed**: 1,000 elements each
- **mixed**: first 100 elements unkeyed, remaining 900 keyed (90% key ratio)

### Results

| Function | unkeyed | keyed | mixed |
|---|---|---|---|
| `doOpCompositeLitCheckFirst` | 1,969 ns | 4,225 ns | panic |
| `doOpCompositeLitSinglePass` | 1,978 ns | 4,295 ns | 4,124 ns |
| `doOpCompositeLitPrescan` | 2,857 ns | 4,035 ns | 3,871 ns |

---

### Recommendation

Given that unkeyed slices are more common in practice, `doOpCompositeLitSinglePass` appears to be the better fit: it handles mixed cases without any performance regression on the unkeyed path, unlike `doOpCompositeLitPrescan` which incurs overhead from the full pre-scan traversal.

If `doOpCompositeLitSinglePass` is adopted, `OpSliceLit` would no longer be used.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 3]
```diff
@@ -341,6 +341,179 @@ func (m *Machine) doOpTypeAssert2() {
 	}
 }
 
+func (m *Machine) doOpCompositeLitCheckFirst() {
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go]: these methods are only used for benchmarking; they should either be kept in test files or, most likely, not kept at all, as they've mostly been used for an investigation and are not useful as on-going benchmarks. the investigation results can be kept and linked to, while keeping the code we have in the codebase lean

**Review by thehowl — CHANGES_REQUESTED** [review]
current implementation has a problem, and arises here:

https://github.com/gnolang/gno/blob/af90c41dc51ad69a0c11504605839db2efb42486/gnovm/pkg/gnolang/op_expressions.go#L359-L381

whether to go with OpSliceLit1 or opSliceLit2 is decided by the first item, so a case like `[]int{1, 2, 3, 11: 123}` still goes through OpSliceLit1.

We should probably make the code more similar to how array values are handled, which already supported mixed keyed/unkeyed elements, and don't have a distinction between OpSliceLit1 and OpSliceLit2

**Comment by aronpark1007** [comment]
The code you reviewed is the pre-fix version. To avoid confusion, I have removed the
  test functions that verified the old panic behavior.                                                                                                                                                                     
   
  Modified to always handle slice composite literals using `OpSliceLit2`, which already                                                                                                                                    
  supports unkeyed, keyed, and mixed keyed elements — eliminating the need for the
  separate `OpSliceLit` path.                                                                                                                                                                                              
                                                                                                                                                                                                                           
  Note: `doOpSliceLit()` and its dispatch case in the machine switch are now dead code                                                                                                                                     
  since `PushOp(OpSliceLit)` is no longer called. I can remove them in this PR if preferred. 

```go
case *SliceType:
    m.PushOp(OpSliceLit2)
    // evaluate item values; also evaluate key if present
    for i := len(x.Elts) - 1; 0 <= i; i-- {
        m.PushExpr(x.Elts[i].Value)
        m.PushOp(OpEval)
        if x.Elts[i].Key != nil {
            m.PushExpr(x.Elts[i].Key)
            m.PushOp(OpEval)
        }
    }
```

**Comment by thehowl** [comment]
yes please, remove the deadcode.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 4]
```diff
@@ -518,29 +505,52 @@ func (m *Machine) doOpSliceLit() {
 
 func (m *Machine) doOpSliceLit2() {
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go]: Please make this code consistent with doOpArrayLit

File: gnovm/tests/files/slice7.gno
Comment (thehowl) [inline gnovm/tests/files/slice7.gno:1]: Please add more tests, including for `[]int{1, 2, 3, 11: 123}` and other edge cases

Comment (aronpark1007) [reply gnovm/tests/files/slice7.gno:1]: Done! Added slice8.gno with additional edge cases including []int{1, 2, 3, 11: 123} and others.

File: gnovm/pkg/gnolang/op_expressions.go (hunk 4 above)
Comment (aronpark1007) [reply gnovm/pkg/gnolang/op_expressions.go]: Done. Refactored doOpSliceLit2 to match doOpArrayLit: keys are now read directly from the AST instead of being evaluated onto the stack at runtime.

**Comment by aronpark1007** [comment]
While working on this, I noticed that non-const keys in slice/array composite literals were not being consistently rejected by the preprocessor. Rather than adding per-type blacklist checks, I replaced the scattered guards with a single whitelist in `CompositeLitExpr TRANS_LEAVE`: any non-nil key that isn't a `*ConstExpr` after preprocessing now panics.

`composite15.gno` and `composite16.gno` have been updated accordingly. Note that the error location now points to the composite literal span rather than the individual key.

For reference, Go also rejects this at compile time: https://go.dev/play/p/d1Ez1UgoUi7

---

Let me know if you'd prefer to keep the per-type blacklist approach, or split this into a separate PR.

**Comment by aronpark1007** [comment]
@thehowl 
Removed the unused `doOpSliceLit` and renamed `doOpSliceLit2` → `doOpSliceLit` accordingly. Let me know if you'd prefer to keep the `2` suffix.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 5]
```diff
@@ -491,78 +474,60 @@ func (m *Machine) doOpArrayLit() {
 	})
 }
 
+
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go:477]: This stray blank line is what's failing both `main / lint` (gofmt/goimports) and `main / build` (the `make generate` diff check). `gofumpt -w` fixes it.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 6]
```diff
@@ -491,78 +474,60 @@ func (m *Machine) doOpArrayLit() {
...
-		idx := itv.ConvertGetInt()
-		if idx > maxVal {
-			maxVal = idx
+		if idx > maxIdx {
+			maxIdx = idx
 		}
 	}
-	m.incrCPU(OpCPUSlopeSliceLit2 * (maxVal + 1))
+	length := maxIdx + 1
+	m.incrCPU(OpCPUSlopeSliceLit * length)
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go:497]: `[]int{9223372036854775806: 1, 2}` is valid Go (passes go/types — verified), but here `length = maxIdx + 1` overflows int64 to negative, and `incrCPU` then runs with negative cycles **before** `NewListArray`'s negative-length guard. On chain, `basicGasMeter.ConsumeGas` Go-panics with `"gas must not be negative"` (tm2/pkg/store/types/gas.go:194) instead of producing a clean VM exception. (Unmetered, it degrades gracefully to `len out of range` via alloc.go:388.)

Not exploitable — the tx aborts either way and consumed gas never decreases — and the single-key variant pre-existed in `doOpSliceLit2`, but since this is being rewritten anyway, add before the `incrCPU` call:

```go
if length < 0 {
	panic(&Exception{Value: typedString("len out of range")})
}
```

A single guard after the first pass suffices: any `idx++` overflow implies `maxIdx == MaxInt64`.

File: gnovm/pkg/gnolang/op_expressions.go [hunk 7]
```diff
@@ -491,78 +474,60 @@ func (m *Machine) doOpArrayLit() {
...
+	vs := m.PopValues(ne)
+	// second pass: fill values using AST keys.
+	idx = -1
+	for i, elt := range x.Elts {
+		if kx := elt.Key; kx != nil {
+			idx = kx.(*ConstExpr).ConvertGetInt()
+		} else {
+			idx++
+		}
 		if es[idx].IsDefined() {
```
Comment (thehowl) [inline gnovm/pkg/gnolang/op_expressions.go:510]: `es[idx].IsDefined()` misses duplicates whose first value is a nil interface: `[]interface{}{nil, 0: 1}` silently overwrites and prints `1 1` instead of panicking "duplicate index" (verified in no-typecheck execution; go/types catches it on chain, so this path is backstop-only). `doOpArrayLit` uses a `set []bool` and catches this correctly — since the goal is consistency with `doOpArrayLit`, use the same approach here. It also removes the subtle reliance on `IsDefined` semantics.

File: gnovm/pkg/gnolang/machine.go [hunk 8]
```diff
@@ -1328,8 +1327,7 @@ const (
 	OpCPUStaticTypeOf    = 520 // XXX arbitrary
 	OpCPUCompositeLit    = 76
 	OpCPUArrayLit        = 292 // base from fit; per-element added in handler
-	OpCPUSliceLit        = 342 // base from fit; per-element added in handler
-	OpCPUSliceLit2       = 966 // base from fit; per-alloc-size added in handler
+	OpCPUSliceLit        = 966 // base from fit; per-alloc-size added in handler
```
Comment (thehowl) [inline gnovm/pkg/gnolang/machine.go:1330]: These constants are now stale: 966/31 are the old `OpSliceLit2` fitted values, now applied to *all* slice literals — including the overwhelmingly common dense case that used to cost 342 + 28/elem, i.e. a ~2.8× base overcharge. Nothing was fitted for the new unified handler (two AST passes + dup checks + zero-fill scan). Recommend re-running calibration with the updated `bench_ops_test.go` benchmarks — or, if a conservative overcharge is intended, say so in the comment instead of "base from fit". Get assistance from claude to get a scaling factor from the base measurements and a new adequate measurement for OpCPUSliceLit.

(Charging the slope per alloc length rather than per element is correct for sparse literals, and `incrCPU` before allocation preserves the DoS ordering — 👍.)

File: gnovm/tests/files/slice8.gno [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package main
```
Comment (thehowl) [inline gnovm/tests/files/slice8.gno:1]: Coverage of the happy paths is good now. Suggest also adding:
- a duplicate-index filetest (`// TypeCheckError:` + `// Error:` directives) — the runtime dup-panic path currently has no coverage;
- a const-identifier-key test (`const k = 3; []int{k: 1, 2}`) — exercises the modified NameExpr preprocess path (works, verified locally, but nothing in the tree covers it);
- the index-overflow case (`[]int{9223372036854775806: 1, 2}`), once guarded.

**Review by davd-gzl — CHANGES_REQUESTED** [review]
Went through this independently at 3c904b4 and agree with @thehowl's comments, nothing material to add from my side. Ping me when it's updated and I'll take another pass.
